# Python As a Dynamically Typed Language

## Introduction

Python is a dynamically typed programming language, which means it handles variable assignments differently compared to statically typed languages like C, C++, and Java. Unlike these languages, Python does not require explicit variable declarations; instead, it relies on assignment statements.

## Static Typing vs. Dynamic Typing

In statically typed languages like C, C++, or Java, when you declare a variable, you set aside a specific area of memory for holding values of a predefined data type. This memory allocation is interpreted strictly according to the data type of the variable. For example, if you declare an integer variable, the allocated memory will be treated as an integer, and attempting to store any other data type in it will result in a compile-time error. This rigid type checking ensures that variables are used consistently with their declared types.

## Dynamic Typing in Python

Python, on the other hand, is dynamically typed. In Python, there is no need to explicitly declare a variable's data type. Instead, Python stores values at specific memory locations and binds variable names to these memory containers. It doesn't determine the variable's type until the code is executed at runtime.

This dynamic typing approach has several implications:

1. No Explicit Declaration: In Python, you do not need to declare the type of a variable before assigning a value to it. You can simply assign a value, and Python will infer the variable's type based on the assigned value.

2. Type Flexibility: Python variables can change their type during execution. For example, you can assign an integer value to a variable and later reassign it to hold a string, and Python will adapt to these changes without any compile-time errors.

3. Run-Time Type Checking: Python checks variable types at runtime, not at compile time. It allows you to mix different data types in your code, and any type-related errors will only be discovered when the code is executed.

In [ ]:
# Dynamic typing in action
x = 42      # x is an integer
x = "Hello" # x is now a string

# No compile-time errors, as Python adapts to the changing type of x

In summary, Python's dynamic typing simplifies variable assignment and allows for greater flexibility in handling data types, as it determines variable types at runtime rather than compile time. This flexibility makes Python a versatile and expressive language for a wide range of programming tasks.

# Python as an Interpreter Language

Python is often categorized as an interpreted programming language. Unlike compiled languages such as C or C++, where source code is translated into machine code before execution, Python source code is directly executed by an interpreter. This article explores the characteristics and advantages of Python as an interpreted language.



# 1. VARIABLES ASSIGNMENT

### Variable Declaration

In Python, you can declare a variable and assign a value to it.


In [44]:
# Variable Declaration and Assignment
x = 10
y = 5.5
name = "John"

**1. Variable Declaration and Assignment:**

* In Python, variables are used to store data or values.
* `x`, `y`, and `name` are examples of variable names.
* `=` is the assignment operator, used to assign a value to a variable.
  
**2. Assigning Values to Variables:**

* `x` is assigned the value `10`. Here, `x` is an integer variable storing the value `10`.
* `y` is assigned the value `5.5`. `y` is a floating-point variable storing a decimal number.
* `name` is assigned the string value `"John"`. `name` is a string variable storing a sequence of characters.
  
So, after executing these lines, you have:

* `x` storing the integer `10`.
* `y` storing the floating-point number `5.5`.
* `name` storing the string `"John"`.

### Variable Types

Python is dynamically typed, meaning you don't need to explicitly declare the data type.

In [ ]:
# Variable Types
a = 10       # Integer
b = 3.14     # Float
c = "Hello"  # String

**1. Variable Types:**

* In Python, variables can hold different types of data.
* `a`, `b`, and `c` are examples of variable names.

**2. Assigning Values to Variables:**

* `a` is assigned the value `10`. `a` is an integer variable, which stores whole numbers without decimal points.
* `b` is assigned the value `3.14`. `b` is a float variable, which stores decimal or floating-point numbers.
* `c` is assigned the string value `"Hello"`. `c` is a string variable, which stores sequences of characters.

### Variable Naming Rules

* Variable names can contain letters, numbers, and underscores.
* They cannot start with a number.
* Case-sensitive (myVar and myvar are different).

In [2]:
# Variable Naming
my_var = 20
anotherVar = 30

In [45]:
# Define a variable named "portfolio_1" and assign a number (integer) to it
# integer is a whole number (no decimals) that could be positive or negative

portfolio_1 = 10.5

In [46]:
# Let's view "portfolio_1"
portfolio_1

10.5

In [48]:
# Define a variable named "portfolio_2" and assign a number (float) to it
# Float are real numbers with a decimal point dividing the integer and fractional parts

portfolio_2 = 500.8

In [49]:
portfolio_2

500.8

In [50]:
x = 2000.44
x

2000.44

In [51]:
x = 10

type_of_x = type(x)

print(type_of_x)


<class 'int'>


In this case, `x` is assigned the value `10`, which is an integer. When you run `type(x)`, it will return `<class 'int'>`, indicating that the variable `x` is of type integer.

So, if you run `type(x)` on the variable `x`, it will output `<class 'int'>`, showing that `x` is an integer.

In [52]:
# Let's overwrite "portfolio_2" (assume your portfolio value increased)
portfolio_2 = 3000.5

In [53]:
# Notice that "portfolio_2" will only contain the most recent value
portfolio_2

3000.5

In [54]:
# Get the type of "portfolio_1" which is integer
# integer is a whole number (no decimals) that could be positive or negative
type(portfolio_1)

float

In [55]:
# Get the type of "portfolio_2" which is float
# Float are real numbers with a decimal point dividing the integer and fractional parts
type(portfolio_2)

float

**MINI CHALLENGE #1:**
- **We defined a variable AMZN (Ticker symbol for Amazon stock) and we assigned these 3 values listed below to it. Without executing any code cells, what will these lines of code generate?**
- **Verify your answer by executing the code cells**



```
AMZN = 3324
AMZN = 3400
AMZN = 3530
AMZN
```


**MINI CHALLENGE #2:**
- **We defined a variable AAPL (Ticker symbol for Apple stock) and we assigned these 2 values listed below to it. Without executing any code cells, what will these lines of code generate?**
- **Verify your answer by executing the code cells**



```
AAPL = 112
AAPL = 115.3
type(AAPL)
```


# 2. MATH OPERATIONS

In [ ]:
# Define a variable named AAPL and initialize it with 5 indicating that we own 5 stocks
# Let's assume that we bought an additional stock so now we have 6 stocks instead of 5
# We can increment AAPL by 1 stock as follows:
AAPL = 100
AAPL = AAPL + 10
AAPL

110

`AAPL = 100`: Here, you are assigning the value `100` to the variable `AAPL`. This line initializes `AAPL` with the value `100`.

`AAPL = AAPL + 10`: In this line, you are updating the value of `AAPL`. `AAPL + 10` calculates the sum of the current value of `AAPL` (which is `100`) and `10`. The result (`110`) is then assigned back to the variable `AAPL`. So, after this line, `AAPL` holds the value `110`.

`AAPL`: If you print the value of `AAPL` after the above operations, it will output `110`. This is the final value of the variable AAPL after the addition.

In [ ]:
AAPL = 100
AAPL -= 20
AAPL

80

`AAPL = 100`: Here, you are assigning the value `100` to the variable `AAPL`. This line initializes `AAPL` with the value `100`.

`AAPL -= 20`: In this line, you are updating the value of `AAPL` by subtracting `20` from its current value. The expression `AAPL -= 20` is a shorthand for `AAPL = AAPL - 20`. So, it subtracts `20` from the current value of `AAPL` (which is `100`) and assigns the result (`80`) back to the variable `AAPL`.

`AAPL`: If you print the value of `AAPL` after the above operations, it will output `80`. This is the final value of the variable `AAPL` after the subtraction.

In [ ]:
# Alternatively, we can increment the variable AAPL by 1 as follows:
AAPL = 10
AAPL += 10
AAPL

20

In [ ]:
# Let's assume that the price of AAPL stock is $260 and we currently have 5 AAPL stocks in our portfolio
# We can calculate the total portfolio value (account balance) as follows:
AAPL_count = 20
AAPL_price = 2000
account_balance = AAPL_count * AAPL_price
account_balance

40000

`AAPL_count = 20`: Here, you are assigning the value `20` to the variable `AAPL_count`. This line initializes `AAPL_count` with the quantity of AAPL (Apple Inc.) stocks you have, which is `20`.

`AAPL_price = 2000`: This line assigns the value `2000` to the variable `AAPL_price`. It represents the price of one share of `AAPL` stock, which is assumed to be $2000 in this example.

`account_balance = AAPL_count * AAPL_price`: Here, you calculate the total account balance by multiplying the number of `AAPL` stocks (`AAPL_count`) with the price per stock (`AAPL_price`). The result (`20 * 2000 = 40000`) is assigned to the variable `account_balance`.

`account_balance`: If you print the value of `account_balance` after the above operations, it will output `40000`. This is the final account balance calculated based on the number of `AAPL` stocks and their price per share.

In [ ]:
# Let's assume you have $20000 USD in our account
# We want to buy AMZN stocks using the total amount
# AMZN stock is priced at $3116 each

account_balance = 20000
AMZN_price = 3116

AMZN_count = account_balance//AMZN_price
AMZN_count

6

`account_balance = 20000`: This line initializes the variable `account_balance` with a value of `20000`, representing the amount of money available in the account.

`AMZN_price = 3116`: Here, you set the variable `AMZN_price` to 3116, which represents the price of one share of AMZN (Amazon.com Inc.) stock.

`AMZN_count = account_balance // AMZN_price`: This line calculates the maximum number of whole shares of `AMZN` stock that can be bought with the given `account_balance`. The `//` operator performs integer division, giving you the maximum count of stocks without considering fractional shares.

`AMZN_count`: If you print the value of `AMZN_count` after the above operations, it will output the calculated count of whole shares of AMZN stock that can be purchased with the given account balance.

In [ ]:
# % returns the remainder

AMZN_count = account_balance % AMZN_price 
AMZN_count

1304

**`AMZN_count = account_balance % AMZN_price`**: This line calculates the remainder when the `account_balance` is divided by the `AMZN_price`. The `%` operator is the modulus operator, and it returns the remainder of the division.

**`**AMZN_count**`**: If you print the value of `AMZN_count` after the above operation, it will output the remainder, which represents the amount of money left after buying as many whole shares of AMZN stock as possible with the given account balance. This is essentially the "change" or "remainder" in terms of the stock price.

In [ ]:
6 % 4

2

In [ ]:
# Divide the account balance by Amazon stock price and place the answer in units
AMZN_price = 5000
divide = account_balance/AMZN_price
divide

10.0

**MINI CHALLENGE #3:**
- **Write a code that calculates the total value of a portfolio assuming that you own 15 Google stocks (Sticker Symbol: GOOG) priced at 1500 each**


**MINI CHALLENGE #4:**
- **Write a code that takes in Google (GOOG) stock prices at two days and calculate the return:**
  - **GOOG price on day 1 = \$260**
  - **GOOG price on day 2 = \$280**

# 3. ORDER OF OPERATIONS (PRECEDENCE)

Order of Operations, also known as precedence, refers to the rules that dictate the sequence in which different operations are performed in a mathematical expression. In Python, as in many programming languages, understanding and following the order of operations is crucial for correctly evaluating expressions.

The order of operations in Python is as follows:

1. **Parentheses** (): Expressions inside parentheses are evaluated first. If there are nested parentheses, the innermost expressions are evaluated first.

2. **Exponents** **: Exponentiation is the next operation. This involves raising a number to a power.

3. **Multiplication** *, **Division** /, **Floor Division** //, **Modulus** %: These operations are performed from left to right.

4. **Addition** +, **Subtraction** -: Addition and subtraction are performed from left to right.

When an expression involves multiple operators, Python follows these rules to determine the correct order of evaluation. For example:

In [7]:
# add 3 and 4 and then multiply the answer by 5
# Note that parantheses have the highest precedence
(2*5) + 5

15

In [8]:
# Multiplication has higher precedence compared to addition
1 * 2 + 3

5

In [9]:
# Division has higher precedence compared to addition
16 / 4 + 25 / 5

9.0

**MINI CHALLENGE #5:**
- **Assume that you own 15 shares of AAPL stock priced at \$150 a share. Apple announced the launch of 5 new products with exceptional features so the stock price increased to $170 a share. Calculate the total profit.**






In [10]:
AAPL_count = 15
initial_price = 150
final_price = 170

Total_profit = (final_price - initial_price) * AAPL_count
Total_profit

300

In [11]:
AAPL_count = 15
AAPL_unitprice_old = 150
AAPL_unitprice_new = 170
Total_profit = (AAPL_unitprice_new * AAPL_count) - (AAPL_unitprice_old * AAPL_count)
Total_profit

300

In [12]:
x = 10
y = 15

z = abs(x - y) * (x + y)
z

125

**MINI CHALLENGE #6:**
- **Write a code that performs the following mathematical operation: z = |x − y| * (x + y)**

In [13]:
x = 2
y = 4
z = abs(x-y) * (x+y)
z

12

In [14]:
x = -15.8373525

x = abs(round(x))
x
f'{x}'

'16'

# 4. PRINT OPERATION

In [15]:
print('Samurai Music City')

Samurai Music City


In [16]:
# Print function is used to print elements on the screen
# Define a string x
# A string in Python is a sequence of characters
# String in python are surrounded by single or double quotation marks

Ekene = 'data scientist'

In [17]:
print(Ekene)

data scientist


In [18]:
type(Ekene)

str

In [19]:
# Obtain the data type for 'x'
type(Ekene)

str

In [20]:
# Print x
print(x)

16


In [21]:
print('The name of my institute is Apple')

The name of my institute is Apple


In [22]:
# Define a string and put 'Apple Inc.' in it
company_name = 'Apple'

In [23]:
# How to use format method in printing

print('The name of my institute is {}'.format(company_name))

The name of my institute is Apple


In [24]:
print('My salary at Microsoft is 1000000')

My salary at Microsoft is 1000000


In [25]:
salary = 1000000000

In [26]:
print('My salary at {} is {}'.format(company_name, salary))

My salary at Apple is 1000000000


In [27]:
# The format() method formats the specified value and insert it in the placeholder.
# The placeholder is defined using curly braces: {}
print("Ticker name for {} is AAPL".format(company_name))

Ticker name for Apple is AAPL


In [28]:
shares = 30

In [29]:
# You can place two or more placeholders with print function
print("I own {} shares of {}".format(shares, company_name))

I own 30 shares of Apple


In [30]:
print(f'I own {shares} shares of {company_name}')

I own 30 shares of Apple


In [31]:
name = 'David'
age = 30
school = 'Gomycode'
country = 'Nigeria'
name = 'Haruna'

In [32]:
print(country)

Nigeria


In [34]:
# Using the fstring method

print(f'My name is {name}, my age is {age}. The name of my school is {school}. I am from {country} ')

My name is Haruna, my age is 30. The name of my school is Gomycode. I am from Nigeria 


In [35]:
# Using the format method

print('My name is {}, my age is {}. The name of my school is {}. I am from {}'.format(name, age, school, country))

My name is Haruna, my age is 30. The name of my school is Gomycode. I am from Nigeria


In [36]:
name = 'David'
age = 21.0

print(f'Name is {name}, and the age is {age}')

Name is David, and the age is 21.0


**MINI CHALLENGE #7:**
- **Write a code that print out the above statement along with AAPL ticker**
- **i.e.: I own 30 shares of Apple Inc.(AAPL)**

In [ ]:
shares = 50

company = 'Microsoft'

print('I own {} shares of {}'.format(shares, company))

I own 50 shares of Microsoft


# 5. GET USER INPUT

In [37]:
name = input('What is your name? ')

In [38]:
name

'Samuel'

In [39]:
print(f'My name is {name}')

My name is Samuel


In [40]:
number1 = input('Provide the first number: ')
number2 = input('Provide the second number')

num = number1 + number2
num

'22'

In [41]:
number1 = int(input('Provide the first number: '))
number2 = int(input('Provide the second number'))

num = number1 + number2
num

4

In [42]:
# input is a built-in function in python
# Get bank client name as an input and print it out on the screen
name = input('Hello, what is your name ')
food = input('What did you have for breakfast ')
print('Hello my name is {}, and I had {}.'.format(name, food))

Hello my name is Samuel, and I had Beans.


In [ ]:
# Obtain bank client data such as name, age and assets and print them all out on the screen
name = input("Welcome to the bank, please enter your name: ")
age = input("Enter your age: ")
networth = input("Enter your total networth (assets - liabilities): ")
print("My name is {}, I am {} years old, and my estimated networth is = {}".format(name, age, networth))

Welcome to the bank, please enter your name: Tomisin
Enter your age: 43
Enter your total networth (assets - liabilities): 500 billion dollars
My name is Tomisin, I am 43 years old, and my estimated networth is = 500 billion dollars


In [ ]:
# Obtain user inputs, perform mathematical operation, and print the results
# Note: Are you getting an error? Move to the next code cell
x = input("Enter the price of AAPL stock today: ")
x = int(x)
y = input("Enter the number of APPLE stocks that you want to buy: ")
y = int(y)
z = x * y
print('The total funds required to buy {} shares of AAPL stock at ${} is ${}'.format(y, x, z))

Enter the price of AAPL stock today: 2000
Enter the number of APPLE stocks that you want to buy: 10
The total funds required to buy 10 shares of AAPL stock at $2000 is $20000


In [ ]:
# Convert from string datatype to integer datatype prior to performing addition

x = input("Enter the price of AAPL stock today: ")
x = int(x)
y = input("Enter the number of APPLE stocks that you want to buy: ")
y = int(y)
print('The total funds required to buy {} shares of AAPL stock at {} dollars is {}'.format(y, x, x*y))

Enter the price of AAPL stock today: 500
Enter the number of APPLE stocks that you want to buy: 30
The total funds required to buy 30 shares of AAPL stock at 500 dollars is 15000


Enter the name of the stock that you want to buy: Apple
Enter the number of stocks that you want to buy: 10
Enter the price of the stock that you want to buy: 200
The total funds required to buy 10 Apple stocks is 2000.


**MINI CHALLENGE #8:**
- **Write a code that takes in the name of the stock, price at which it is selling, the number of stocks that you want to own and prints out the total funds required to buy this stock. Find a sample expected output below:**

  - Enter the price of the stock you want to buy: 3000  
  - Enter the number of stocks that you want to buy: 5
  - Enter the name of the stock that you want to buy: AMZN
  - The total funds required to buy 5 number of AMZN stocks at 3000 is 15000


In [ ]:
10 % 3

1

In [ ]:
price = input('Enter the price of the stock you want to buy: ')
price = int(price)
number = input('Enter the number of stocks that you want to buy: ')
number = int(number)
name = input('Enter the name of the stock that you want to buy: ')
print('The total funds required to buy {} number of {} stocks at {} is {}'.format(number, name, price, number*price))

Enter the price of the stock you want to buy: 1500
Enter the number of stocks that you want to buy: 100
Enter the name of the stock that you want to buy: Amazon
The total funds required to buy 100 number of Amazon stocks at 1500 is 150000


# EXCELLENT JOB

# MINI CHALLENGES SOLUTIONS

**MINI CHALLENGE #1:**
- **We defined a variable AMZN (Ticker symbol for Amazon stock) and we assigned these 3 values listed below to it. Without executing any code cells, what will these lines of code generate?**
- **Verify your answer by executing the code cells**



```
AMZN = 3324
AMZN = 3400
AMZN = 3530
AMZN
```


In [ ]:
# The output of this code is 3530
# Initially we put 3324 in AMZN, then we overwrite it by placing 3400 in AMZN, and then 3530 in AMZN
AMZN = 3324
AMZN = 3400
AMZN = 3530
AMZN

3530

**MINI CHALLENGE #2:**
- **We defined a variable AAPL (Ticker symbol for Apple stock) and we assigned these 2 values listed below to it. Without executing any code cells, what will these lines of code generate?**
- **Verify your answer by executing the code cells**



```
AAPL = 112
AAPL = 115.3
type(AAPL)
```

In [ ]:
AAPL = 112
AAPL = 115.3
type(AAPL)

float

**MINI CHALLENGE #3 SOLUTION:**
- **Write a code that calculates the total value of a portfolio assuming that you own 15 Google stocks (Sticker Symbol: GOOG) priced at 1500 each**


In [ ]:
price  = 1500
x  = 15
amount = x * price
amount

22500

**MINI CHALLENGE #4 SOLUTION:**
- **Write a code that takes in Google (GOOG) stock prices at two days and calculate the return:**
  - **GOOG price on day 1 = \$260**
  - **GOOG price on day 2 = \$280**

In [ ]:
GOOG_price_1 = 260
GOOG_price_2 = 280

price_diff = GOOG_price_2 - GOOG_price_1
percentage_change = price_diff / GOOG_price_1 * 100

percentage_change

7.6923076923076925

**MINI CHALLENGE #5 SOLUTION:**
- **Assume that you own 15 shares of AAPL stock priced at \$150 a share. Apple announced the launch of 5 new products with exceptional features so the stock price increased to $170 a share. Calculate the total profit.**


In [ ]:
AAPL_new = 170
AAPL_old = 150
num_shares = 15
profit = (AAPL_new - AAPL_old) * num_shares
profit

300

**MINI CHALLENGE #6 SOLUTION:**
- **Write a code that performs the following mathematical operation: z = |x − y| * (x + y)**

In [ ]:
x = 4
y = 2
z = abs(x-y) * (x+y)

**MINI CHALLENGE #7 SOLUTION:**
- **Write a code that print out the above statement along with AAPL ticker**
- **i.e.: I own 30 shares of Apple Inc.(AAPL)**

In [ ]:
ticker = 'AAPL'
print("I own {} shares of {}({})".format(shares, company_name, ticker))

I own 30 shares of Apple Inc.(AAPL)


**MINI CHALLENGE #8 SOLUTION:**
- **Write a code that takes in the name of the stock, price at which it is selling, the number of stocks that you want to own and prints out the total funds required to buy this stock. Find a sample expected output below:**

  - Enter the price of the stock you want to buy: 3000  
  - Enter the number of stocks that you want to buy: 5
  - Enter the name of the stock that you want to buy: AMZN
  - The total funds required to buy 5 number of AMZN stocks at 3000 is 15000



In [ ]:
x = input("Enter the price of the stock you want to buy: ")
x = int(x)
y = input("Enter the number of stocks that you want to buy: ")
y = int(y)
z = input("Enter the name of the stock that you want to buy: ")
print('The total funds required to buy {} number of {} stocks at {} is {}'.format(y,z,x, x*y))

Enter the price of the stock you want to buy: 3000
Enter the number of stocks that you want to buy: 5
Enter the name of the stock that you want to buy: AMZN
The total funds required to buy 5 number of AMZN stocks at 3000 is 15000


Objective
Congratulations, you've got a job at Python Pizza. Your first job is to build an automatic pizza order program. Based on a user's order, work out their final bill.

 ``` Small Pizza: $15 ``` ```

``` Medium Pizza: $20 ```

``` Large Pizza: $25 ```

If they want to add pepperoni :

``` Pepperoni for Small Pizza: +$2 ```

``` Pepperoni for Medium or Large Pizza: +$3 ```

If they want cheese :

 ``` Extra cheese for any size pizza: + $1 ```

Instruction
Open a new file in your Python development environment or IDE.
Begin by typing the line 'Welcome to Python Pizza Deliveries' which will display a welcome message when the code is run.
Add a line that prompts the user for the size of the pizza they want using the input() function and stores the user's input in a variable named 'size'.
Add a line that prompts the user if they want pepperoni on their pizza using the input() function and stores the user's input in a variable named 'add_pepperoni'.
Add a line that prompts the user if they want extra cheese on their pizza using the input() function and stores the user's input in a variable named 'extra_cheese'.
Check your code for any syntax errors or mistakes.
Run the code and test it by providing inputs for size, pepperoni, and extra cheese.
Example Input

size = "L"

add_pepperoni = "Y"

extra_cheese = "N"

Example Output

Your final bill is: $28.

When you hit **run**, this is what should happen

In [43]:
# Welcome message
print("Welcome to Python Pizza Deliveries!")

# Get user input for pizza size
size = input("What size pizza do you want? S, M, or L: ").upper()

# Get user input for pepperoni
add_pepperoni = input("Do you want to add pepperoni? Y or N: ").upper()

# Get user input for extra cheese
extra_cheese = input("Do you want extra cheese? Y or N: ").upper()

# Set the initial cost based on pizza size
cost = 0

if size == "S":
    cost += 15
elif size == "M":
    cost += 20
elif size == "L":
    cost += 25
else:
    print("Invalid pizza size. Please choose S, M, or L.")

# Add pepperoni cost if requested
if add_pepperoni == "Y":
    if size == "S" or size == "M":
        cost += 2
    else:
        cost += 3

# Add extra cheese cost if requested
if extra_cheese == "Y":
    cost += 1

# Display the final bill
print(f"Your final bill is: ${cost}.")


Welcome to Python Pizza Deliveries!
Your final bill is: $18.
